### PR TITLE
feat(torch): ocr model training and inference

### PR DESCRIPTION
**Pull request by Bycob**
TODO:
- [x] export backbone with feature extraction example
- [x] create service example call
- [x] train example call
- [x] predict example call
- [ ] document template params
- [ ] bidirectional LSTM (future PR?)

## Export backbone:

```bash
cd tools/torch/
# Display extractable layers
./trace_torchvision.py resnet50 --print-extract

# Extract specific layer
./trace_torchvision.py -v resnet50 --extract layer2
```

## Create service example

```bash
curl -X POST "http://10.10.77.105:8500/services/ocr_test" -d '{
    "mllib": "torch",
    "description": "OCR example",
    "type": "supervised",
    "parameters": {
        "input": {
            "connector": "image",
            "width": 112,
            "height": 32,
            "db": true,
            "ctc": true
        },
        "mllib": {
            "template": "crnn",
            "finetuning": true,
            "gpu": true,
            "gpuid": 0
        },
        "output": {
            "store_config": true
        }
    },
    "model": {
        "repository": "/opt/platform/models/private/ocr_test/"
    }
}'
```
## Train service example
```bash
curl -X POST "http://10.10.77.105:8500/train/" -d '{
    "service": "ocr_test",
    "async": true,
    "parameters": {
        "input": {
            "shuffle": true,
            "db": true,
            "ctc":true
        },
        "mllib": {
            "gpu": true,
            "gpuid": 0,
            "resume": false,
            "mirror": true,
            "geometry":{
                "prob": 0.1,
                "persp_horizontal": true,
                "persp_vertical": true,
                "zoom_in": true,
                "zoom_out": true,
                "pad_mode": "constant"
            },
            "noise": {
                "prob": 0.1
            },
            "distort": {
                "prob": 0.1
            },
            "net": {
                "batch_size": 128,
                "test_batch_size": 32
            },
            "solver": {
                "test_initialization": false,
                "iterations": 1000000,
                "test_interval": 1000,
                "snapshot": 2000,
                "base_lr": 0.0001,
                "solver_type": "RANGER",
                "sam": true,
                "swa": false,
                "iter_size": 1
            }
        },
        "output": {
            "measure": [
                "acc"
            ]
        }
    },
    "data": [
        "/opt/platform/data/ocr_dataset/train_shuf.txt",
        "/opt/platform/data/ocr_dataset/test_shuf.txt"
    ]
}'
```
## Predict example

```bash
curl -X POST "http://10.10.77.105:8500/predict" -d '{
    "service": "ocr_test",
    "parameters": {
        "mllib": {
            "gpuid": 0
        },
        "output": {
            "ctc": true,
            "confidence_threshold": 0,
            "blank_label": 0,
            "best": 1
        }
    },
    "data": [
        "/opt/platform/data/imgs/text0.png"
    ]
}'
```

### CODE REVIEW

File: src/backends/torch/native/templates/crnn.hpp [hunk 1]
```diff
@@ -0,0 +1,125 @@
+/**
+ * DeepDetect
+ * Copyright (c) 2020 Jolibrain
+ * Author:  Emmanuel Benazera <emmanuel.benazera@jolibrain.com>
```
Comment (beniz) [inline src/backends/torch/native/templates/crnn.hpp:4]: Fix author

File: src/backends/torch/torchdataset.cc [hunk 2]
```diff
@@ -840,6 +840,59 @@ namespace dd
...
+        = torch::full(1, at::Scalar(int(target.size())),
+                      at::TensorOptions().dtype(torch::kInt64));
+    int i = 0;
+
+    for (auto &c : target)
+      {
+        if (i >= max_ocr_length)
+          {
+            // can happen in test set
+            this->_logger->warn("Sequence \"{}\" is exceding maximum ocr "
```
Comment (beniz) [inline src/backends/torch/torchdataset.cc:860]: "exceeding"

File: src/backends/torch/torchdataset.cc [hunk 3]
```diff
@@ -840,6 +840,59 @@ namespace dd
...
+          }
+        auto it = alphabet.find(c);
+
+        if (it != alphabet.end())
+          {
+            target_tensor[i] = it->second;
+          }
+        else if (!_test)
+          {
+            std::cout << "added " << c << " to alphabet" << std::endl;
```
Comment (beniz) [inline src/backends/torch/torchdataset.cc:873]: use logger instead ?

File: src/backends/torch/torchdataset.cc [hunk 4]
```diff
@@ -840,6 +840,59 @@ namespace dd
...
+            this->_logger->warn(
+                "Character {} in test set but not in train set", c);
+          }
+
+        i++;
+      }
+    static int counter = 0;
+    if (counter++ % 1000 == 0)
+      {
+        std::cout << "Added string " << target
```
Comment (beniz) [inline src/backends/torch/torchdataset.cc:889]: Use logger instead ?

File: src/backends/torch/torchinputconns.cc [hunk 5]
```diff
@@ -491,6 +515,86 @@ namespace dd
                     _test_datasets[i].add_image_image_file(
                         lfile.first, lfile.second, _height, _width);
               }
+            else if (_ctc)
+              {
+                std::vector<std::pair<std::string, std::string>>
+                    lfiles; // labeled files
+                std::vector<std::vector<std::pair<std::string, std::string>>>
+                    tests_lfiles;
+                std::cout << "Reading ctc dataset" << std::endl;
```
Comment (beniz) [inline src/backends/torch/torchinputconns.cc:524]: Logger ?

File: src/backends/torch/torchlib.cc [hunk 6]
```diff
@@ -1467,8 +1487,12 @@ namespace dd
                     output = torch::stack(outputs);
                   }
 
-                if (extract_layer.empty() && _classification && !_timeserie)
+                // XXX: why is (!_timeserie) needed here? Aren't _timeserie and
+                // _classification mutually exclusive?
```
Comment (fantes) [inline src/backends/torch/torchlib.cc:1491]: timeseries could be classified (but  we never had such use case)

**Review by fantes — APPROVED** [review]

File: src/backends/torch/torchlib.cc (hunk 6 above)
Comment (Bycob) [reply src/backends/torch/torchlib.cc:1491]: Will add a comment

**Review by beniz — APPROVED** [review]

**Pull request merged**